numpy/ma/core.py

```diff
@@ -3183,29 +3183,71 @@ def __getitem__(self, indx):
         Return the item described by i, as a masked array.
 
         """
-        dout = self.data[indx]
         # We could directly use ndarray.__getitem__ on self.
         # But then we would have to modify __array_finalize__ to prevent the
         # mask of being reshaped if it hasn't been set up properly yet
         # So it's easier to stick to the current version
+        dout = self.data[indx]
         _mask = self._mask
+
+        def mout_and_scalar_expected(_mask):
+            """
+            Returns the mask for the selected element, and a boolean for whether
+            this is a scalar index.
+
+            The heuristics for this are a little messy when there is no mask
+            """
+            if _mask is nomask:
+                # obviously a scalar
+                if not isinstance(dout, np.ndarray):
+                    return nomask, True
+
+                # object array scalar indexing can return anything
+                elif self.dtype.type is np.object_:
+                    if self.dtype is not dout.dtype:
+                        # dout is an array, but dtypes do not match, so must be
+                        # an element
+                        return nomask, True
+
+                # well-behaved subclass that only returns 0d arrays when
+                # expected - this is not a scalar
+                elif type(self.data).__getitem__ == ndarray.__getitem__:
+                    return nomask, False
+
+                # heuristics have failed
+                # construct a full array, so we can be certain. This is costly
+                _mask = getmaskarray(self)
+
+            # _mask cannot be a subclass, so it tells us whether we should
+            # expect a scalar. It also cannot be of dtype object.
+            mout = _mask[indx]
+            return mout, not isinstance(mout, np.ndarray)
+
+        mout, scalar_expected = mout_and_scalar_expected(_mask)
+
         # Did we extract a single item?
-        if not getattr(dout, 'ndim', False):
+        if scalar_expected:
             # A record
             if isinstance(dout, np.void):
-                mask = _mask[indx]
                 # We should always re-cast to mvoid, otherwise users can
                 # change masks on rows that already have masked values, but not
                 # on rows that have no masked values, which is inconsistent.
-                dout = mvoid(dout, mask=mask, hardmask=self._hardmask)
+                return mvoid(dout, mask=mout, hardmask=self._hardmask)
+
+            # special case introduced in gh-5962
```

Comment:
I still don't see where the `MaskedConstant` is being `__new__`'d from, causing the problem you described below